File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/AnalysisTagResponse.java
package openfoodfacts.github.scrachx.openfood.models;

import java.util.ArrayList;
import java.util.Map;

/**
 * Intermediate class between {@link AnalysisTagsWrapper} and {@link AnalysisTag}
 *
 * @author Rares
 */
public class AnalysisTagResponse {
    private static final String DEFAULT_LANGUAGE = "en";
    private final Map<String, String> namesMap;
    private final Map<String, String> showIngredientsMap;
    private final String uniqueAnalysisTagID;

    /**
     * Constructor.
     *
     * @param uniqueAnalysisTagId
     * @param namesMap
     * @param showIngredientsMap
     */
    public AnalysisTagResponse(String uniqueAnalysisTagId, Map<String, String> namesMap, Map<String, String> showIngredientsMap) {
        this.uniqueAnalysisTagID = uniqueAnalysisTagId;
        this.namesMap = namesMap;
        this.showIngredientsMap = showIngredientsMap;
    }

    /**
     * Converts an AnalysisTagResponse object into a new AnalysisTag object.
     *
     * @return The newly constructed AnalysisTag object.
     */
    public AnalysisTag map() {
        AnalysisTag analysisTag;
        analysisTag = new AnalysisTag(uniqueAnalysisTagID, new ArrayList<>());
        for (Map.Entry<String, String> name : namesMap.entrySet()) {
            String showIngredients = showIngredientsMap.get(name.getKey());
            if (showIngredients == null) {
                showIngredients = showIngredientsMap.get(DEFAULT_LANGUAGE);
            }
            analysisTag.getNames()
                .add(new AnalysisTagName(analysisTag.getTag(), name.getKey(), name.getValue(), showIngredients));
        }

        return analysisTag;
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/analysistag/AnalysisTagResponse.kt
package openfoodfacts.github.scrachx.openfood.models.entities.analysistag

import openfoodfacts.github.scrachx.openfood.models.entities.EntityResponse

/**
 * Intermediate class between [AnalysisTagsWrapper] and [AnalysisTag]
 *
 * @author Rares
 */
class AnalysisTagResponse(
    private val uniqueAnalysisTagID: String,
    private val namesMap: Map<String, String>,
    private val showIngredientsMap: Map<String, String>
) : EntityResponse<AnalysisTag> {
    /**
     * Converts an AnalysisTagResponse object into a new AnalysisTag object.
     *
     * @return The newly constructed AnalysisTag object.
     */
    override fun map(): AnalysisTag {
        val analysisTag = AnalysisTag(uniqueAnalysisTagID, arrayListOf())

        analysisTag.names += namesMap.map { (lc, name) ->
            val showIngredients = showIngredientsMap[lc] ?: showIngredientsMap[DEFAULT_LANGUAGE]
            AnalysisTagName(analysisTag.tag, lc, name, showIngredients)
        }
        return analysisTag
    }

    companion object {
        private const val DEFAULT_LANGUAGE = "en"
    }
}